Update the Get Started documentation

brian-dlee opened the pull request:
As I get myself introduced to Exosphere, I'm updating the documentation to reflect the current implementation of Exosphere as it has evolved over the past few months.

@hugobho

brian-dlee committed:
Ignore Visual Studio Code project files

brian-dlee committed:
Update installation instructions

Change installation instructions from installing via NPM to the manual download and install of binaries from the GitHub release page of the project

brian-dlee committed:
Minor install doc text update

Updated install doc to refer to application as Exosphere instead of exo

brian-dlee committed:
Readd the navigation to installation page

brian-dlee committed:
Update symlink command on installation page

Updated the `ln -s` to `ln -fns` to forcefully update a link if it already exists

brian-dlee committed:
Update exo tool page

Correct code formatting for the "exo create application" command
Add ExoCom prompt question to list of questions to expect when running exo create application

brian-dlee committed:
Update Application Shell page

Update the `application.yml` preview to reflect the content that was created for me when running `exo create application`
Update the description of the config file with information regarding the update I added on the development dependency `exocom`

kevgo commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+.vscode
```
Tool-specific configuration should be in the [global gitignore](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore) on your device, not in every code base you work on.

kevgo commented on website/tutorial/part_1/03_installation.md:
```diff
@@ -11,68 +11,32 @@
 
-The installation instructions below include
-making these infrastructure components available on your system.
+* Download and install the binary executable for your platform from the 
```
Please link to the latest release: https://github.com/Originate/exosphere/releases/latest. This way people always install the latest version.

kevgo commented on website/tutorial/part_1/03_installation.md:
```diff
@@ -11,68 +11,32 @@
+* Download and install the binary executable for your platform from the 
+[GitHub release page](https://github.com/Originate/exosphere/releases)
+* Rename the executable to `exo`
+* Check and update your `PATH` environmental variable to include its install location, if required
```
I would suggest moving the `exo` binary into a directory that is in the path. This is common for installation instructions.

brian-dlee replied:
I've reworked the install instructions, but only have specific installation instructions for Mac. I'm not familiar enough with scripting on Windows to write Windows instruction.

Do you think it's safe to assume a Linux user is able to install without explicit instructions? Especially considering Linux instructions will be nearly identical to the Mac instructions.

kevgo replied:
Don't worry about platforms you are not familiar with. Thanks for doing this for the Mac!

kevgo commented on website/tutorial/part_1/03_installation.md:
```diff
@@ -11,68 +11,32 @@
+the version in the filename and symlinking the desired command name to the executable.
+Using this pattern, old versions are still available in your `PATH`. In the event that you need
+to run an old version of Exosphere it will still be available as `exo-X.Y.Z` unless it is explicitly 
+removed.
```
Is that really a thing right now? Is there a use case for this? Afaik we don't provide support for older versions since we don't have stable releases yet. I would prefer that everybody always uses the latest version. Once we reach a the second stable release (2.0), we can support older stable releases (1.x) longer.

kevgo commented on website/tutorial/part_1/03_installation.md:
```diff
@@ -11,68 +11,32 @@
-making these infrastructure components available on your system.
+* Download and install the binary executable for your platform from the 
+[GitHub release page](https://github.com/Originate/exosphere/releases)
+* Rename the executable to `exo`
```
Please also include that people need a recent version of Git and Docker on their system.

kevgo reviewed (commented):
Thanks for helping with this, @brian-dlee! Some comments below.

brian-dlee committed:
Drop tool specific .gitignore rule

brian-dlee committed:
Update install page release download URL

Changed the URL to point to the latest release only

brian-dlee committed:
Updates install page with external dependencies

Adds a section that provides links to other applications that are required to run Exosphere

brian-dlee committed:
Updates install instructions

Removes text on managing multiple versions
Changes install steps to assert the binary is installed in a location that should already be found in the path

brian-dlee committed:
Updates HTML Server page in tutorial

Corrects the service.yaml file preview
Adds instructions on adding the html-server template to exo
Updates prompt instruction on `exo add service`
Removes the `exo setup` command since it doesn't exist anymore
Updates the output for `exo run`

brian-dlee committed:
Updates some text on the HTML Server page

Moves explanatory text before instructions block
Adds a small line to explain the following code block will be an output preview

brian-dlee commented on website/tutorial/part_2/part_2a/04_html_server.md:
````diff
@@ -110,138 +115,161 @@ Here is the current architecture of our application:
- exocom      received routing setup
- exorun      application ready
-```
+* **exocom** is the command that runs Exosphere applications. It starts the
````
I could use a little assistance here. I'm not sure if replacing exorun with exocom in this text block is actually correct. Is exorun just not used here?

brian-dlee replied:
I think I found an appropriate change to address this. Of course, any and all feedback is welcome.

brian-dlee committed:
Adjusts output description on HTML Server page

Removes references to exorun; There is no visible evidence that `exorun` exists in this example.

brian-dlee committed:
Updates link to Exosphere Architecture page

brian-dlee committed:
Fixes minor typos on the communication page

brian-dlee committed:
Updates to todo service

Added reference to previous html-server service implementation
Adds instruction for including mongodb exoservice template
Updates prompt questions for exo add service
Updates service.yml for todo service based on what came out when I ran it
Removes exo setup command

kevgo commented on website/tutorial/part_1/03_installation.md:
```diff
@@ -11,68 +11,32 @@
 
-### macOS
+Before installing Exosphere, ensure you have the following applications installed:
+ * Git - [https://git-scm.com/]()
```
I'd format the name of the tools as the link, like so:

```markdown
* [Git](https://git-scm.com)
* [Docker](https://www.docker.com)

kevgo commented on website/tutorial/part_1/03_installation.md:
````diff
@@ -11,68 +11,32 @@
 
+In a terminal, execute the following:
+```bash
+curl -L https://github.com/Originate/exosphere/releases/download/vX.Y.Z/exo-darwin-amd64 >/usr/local/bin/exo
````
Not sure how much easier it is to provide a curl command here. Users have to look up the version manually on the website anyways - why not have them download the correct binary right there?

We could tell them to move the binary to `/usr/local/bin` if they don't have a good other place to put it.

kevgo commented on website/tutorial/part_2/part_2a/04_html_server.md:
```diff
@@ -6,66 +6,68 @@
-we'll build it using [ExpressJS](http://expressjs.com).
-Exosphere provides a template for building ExpressJS html servers.
-Let's use it:
+Let's build the first service for our application: the HTML server! If we were
```
Please try to match the existing flow of line breaks. The existing text breaks at logical breaking points like commas, subclauses, and each time a thought chain ends. This has several advantages: it makes it much easier to read the markdown code (okay, that's debatable, but I find it easier), and it limits changes to the respective line their are in. If we followed a constant line length, then just adding one world somewhere potentially changes all lines after it, causing massive diffs where it is hard to spot the actual change. Plus, not everybody has a sophisticated editor that automatically breaks text at constant line lengths, and we don't want to exclude these people. Thanks!  :)

brian-dlee replied:
I agree. 

I think my editor took this page by storm. I caught it in it's antics in the the following page and disabled that behavior. I'll revert all of it's formatting changes in this page.

kevgo commented on website/tutorial/part_2/part_2a/04_html_server.md:
```diff
@@ -110,138 +115,159 @@ Here is the current architecture of our application:
+service.
 
-__~/todo-app/html-server/service.yml__
+**~/todo-app/web-service/service.yml**
```
Why do you change `__` to `**` here? It doesn't really change anything.

brian-dlee replied:
I think this was the formatter. I don't remember consciously changing this nor do I have a good reason if I did. Either way, I will blame the formatter and correct it.

kevgo commented on website/tutorial/part_2/part_2a/04_html_server.md:
```diff
@@ -110,138 +115,159 @@ Here is the current architecture of our application:
+latest: Pulling from library/node
+85b1f47fba49: Pull complete
+ba6bd283713a: Pull complete
+817c8cd48a09: Pull complete
```
This output is too lengthy. Can you please eliminate some repeated lines, like these ones here showing `xxxx: Pull complete`? Maybe keep only 2 to show there can be multiple ones...

kevgo commented on website/tutorial/part_2/part_2a/04_html_server.md:
````diff
@@ -110,138 +115,159 @@ Here is the current architecture of our application:
 ```
 
-You should see output like:
+The Exosphere framework itself is written as a bunch of loosely coupled
````
This isn't correct anymore. Exosphere is now a single monolithic Go binary. Exocom is a different product, and the web service is part of the application.

brian-dlee replied:
Were you implying that I should remove the reference to exocom from the list? I redid this but kept the text referring to exocom. If that isn't desired I will adjust.

kevgo replied:
Let's focus this tutorial on Exosphere and leave Exocom and micro-services out as much as possible. Exosphere works equally well for monoliths. We can use Exocom in the example application, but don't have to go into details on how that application works.

kevgo commented on website/tutorial/part_2/part_2a/04_html_server.md:
```diff
@@ -110,138 +115,159 @@ Here is the current architecture of our application:
 Takeaway:
-> The web server in a microservice application is much simpler than in a monolith,
-> because it only focuses on interacting with the user via HTML.
+
```
Why do you add an empty line here?

kevgo commented on website/tutorial/part_2/part_2a/08_todo_service.md:
```diff
@@ -9,6 +9,12 @@
 
 # Building the Todo service
 
+> Note: This is the second service of the tutorial. 
```
Do we have to say that here? I think we can assume that people follow a tutorial correctly. This could be said on every page.

brian-dlee replied:
The reason I stated this here is because this page relies on concrete changes from previous pages. This page instructs you to select template number 2. If template number 1 has not been added then it makes this page function incorrectly. Basically, I added it to mitigate confusion in case anyone finds their way to this page without following the appropriate steps.

If you still feel it's unnecessary considering this then I have no problem removing it.

kevgo replied:
I think we can assume that people aren't idiots and can follow a tutorial correctly as long as all changes we want them to do are easily visible and clearly spelled out. If there are particular concerns that people will get tripped up here, maybe we can provide a working copy of the code base for this stage somewhere in a folder? That way, if our readers don't get it working, they can look at our code and see what is different.

kevgo reviewed (commented):
Very nice! Please revert the changes to the line breaks, then this is pretty much good to go from my perspective.

brian-dlee committed:
Adjusts dependency links on installation page

brian-dlee committed:
Adjusts the download steps for Mac

Removed the use of the curl command in favor of a non-scripted download procedure

brian-dlee committed:
Updates HTML Server page

Reverts all automated editor formatting from a previous commit
Converts prompt question to lowercase to be consistent with other pages
Reverted web-service to html-server to be consistent with the page name; Previously I changed it to web-service because that was the default option from the template and I was unsure of the consequences of changing it
Removed text describing the Exosphere framework as a series of small tools
Trimmed the output section to only the most relevant information

brian-dlee committed:
Applies minor text change on Installation page

brian-dlee committed:
Updates prompt tables

After reviewing how these tables look to the user I think the clearer the better; I'm moving these tables away from single word lower case prompt values and being more explicit about defaults in the "text you enter" section

brian-dlee committed:
Fixes URL to mongodb exoservice in clone command

brian-dlee committed:
Updates the service.yml format for ToDo service

After finding that dependencies were not placed correctly with Hugo the exoservice was updated and I've added the updated service.yml in the tutorial

brian-dlee committed:
Applies minor text update on Todo Service page

brian-dlee committed:
Updates todo service page emphasized text format

brian-dlee committed:
Updates integration with web server page

Reorder steps slightly to allow easier testing; Updating accepted messages first allows testing of the views before all the steps are complete.
Fixed some semantic issues in the example *.js files
Other minor Markdown fixes
Updated output of exo run

brian-dlee committed:
Updates some outdated text about the html-server

brian-dlee committed:
Removes $ from prompt for consistency

brian-dlee committed:
Fixed bad link on web server integration page
 Please enter the commit message for your changes. Lines starting

kevgo commented on website/tutorial/part_1/03_installation.md:
```diff
@@ -11,67 +11,34 @@
 
-## Windows
+_Note: Depending on the privileges of `/usr/local/bin` on your system 
+some of these steps could require root privileges_
```
I would also add: 
> If you don't want to use root privileges here, please move Exosphere to a different folder in your home directory that is in the $PATH

kevgo replied:
Please move this hint down to the line where it applies (line 40).

kevgo commented on website/tutorial/part_1/03_installation.md:
```diff
@@ -11,67 +11,34 @@
 
-### macOS
+Before installing Exosphere, ensure you have the following applications installed:
+ * [Git](https://git-scm.com/)
```
Please remove the trailing slash from the URL here.

kevgo commented on website/tutorial/part_2/part_2a/04_html_server.md:
```diff
@@ -6,44 +6,47 @@
-tracking user sessions and permissions,
-rendering HTML for the browser, JSON for the REST API,
-sending out emails,
+storing todo items, user accounts, and API tokens in the database, 
```
Sorry to bother you about this again. Your text editor is adding trailing spaces to all lines here. Please configure it to trim trailing spaces on save. There are ways to do that for all common editor types. Google is your friend here. :)

brian-dlee replied:
No sweat. It looks like turning the formatter off altogether caused some issues.

kevgo reviewed (commented):
This is coming along nicely, almost there!  :)

brian-dlee committed:
Merge remote-tracking branch 'origin/master' into bl-update-get-started-documentation

brian-dlee committed:
Dropped trailing slash in install dependencies

brian-dlee committed:
Adjusts Mac installation note

Includes information on how to bypass the use of root privileges
Moves notice down right below the command

brian-dlee committed:
Trims trailing whitespace from HTML Server page

brian-dlee committed:
Drops previous tutorial step reference from Todo Service page

brian-dlee committed:
Wraps long line on installation page
Trims trailing whitespace on installation page

brian-dlee committed:
Trims trailing whitespace on a couple pages

kevgo commented on website/tutorial/part_2/part_2a/02_app_config.md:
```diff
@@ -25,14 +25,20 @@ name: todo-app
 It defines the name, a description, and a version of the application.
-It also defines a section for listing all the services of the application.
-It is empty right now, since our application doesn't contain any services yet.
+It also defines a section for listing all the services of the application and
```
> It also defines a section for listing all the services and dependencies of the application. In our case this includes `exocom`.

kevgo commented on website/tutorial/part_2/part_2a/04_html_server.md:
```diff
@@ -6,12 +6,12 @@
   </tr>
 </table>
 
-
```
Please don't delete this line. I think having two empty lines in front of headers helps separate them visually.

kevgo commented on website/tutorial/part_2/part_2a/04_html_server.md:
```diff
@@ -6,12 +6,12 @@
 the HTML server!
-If we were building our Todo app as a traditional monolith,
+If we were
+building our Todo app as a traditional monolith,
```
Why the line break here?

kevgo commented on website/tutorial/part_2/part_2a/04_html_server.md:
```diff
@@ -35,11 +35,14 @@ than it would be in a traditional monolithic application.
+we will add the HTML server service template and then we will add the new service.
 
 <a class="runMarkdown_consoleWithInputFromTable">
+
```
Why the empty line here?

brian-dlee replied:
The code block doesn't render correctly via Markdown preview without the line break between the anchor tag and the opening of the code block (```).

kevgo replied:
Okay, makes sense. Please let me know when this is ready to be looked at again.

kevgo reviewed (commented):
Once these changes are addressed, this is good to go. Thanks again, Brian!

brian-dlee committed:
Includes minor text update to app config page

brian-dlee committed:
Includes minor whitespace adustments to HTML server page

brian-dlee committed:
Adds some missing newlines on the HTML Server page

brian-dlee committed:
Readds one more missing newline on HTML Server page

brian-dlee commented:
Alright @kevgo, I think I've addressed all the whitespace issues. Take a look.

kevgo commented:
Thanks!! 👍
